### DISCUSSION TITLE
Allow targeting specific cart lines with discount functions

### DISCUSSION THREAD

**Discussion by mariusa** [Shopify/function-examples#199] [Feedback: Discounts APIs]
The same variant ID can be in cart on multiple lines, with different attributes. But Shopify functions output requires to specify a variant ID instead of cart line item id:
```
"targets": [
        {
          "productVariant": {
            "id": "gid://shopify/ProductVariant/1"
          }
        }
      ],
```
To keep compatibility, another target type could be added (eg `cartLineItem`), besides `productVariant`.

Also requested by other partners on forums, ignored by Shopify: https://community.shopify.com/c/shopify-functions/feature-request-add-ability-to-discount-a-specific-cart-line/td-p/1801254

**Comment by realpratiknikam** [comment]
This is really a blocker for attributes based discount.

**Comment by evr-novicell** [comment]
Just found out about this limitation today. Huge issue for Shopify plus clients where discounts might vary from line item to line item due to complex discount rules stores in cart line properties and where cart lines should be targeted by cart line item id.

Here's a full example of the problem:

Product Discount "ProductVariantTarget" accepts the "id" of the target variant. Does that mean we cannot apply discounts to specific cart line items?

- For example, I am selling printed shirts and the customer can enter a text to be printed on the shirt.
- If the message length is less than or equal to 3 characters, the item should get a 50% discount
- Customer adds a product variant with id "123" to the cart, and cart line property "text" is added with value "abcd"
- Customer adds a product variant with id "123" to the cart again and this time cart line property "text" is added with value "ABC"

Cart contains two line items of the same product variant because of the difference in cart line property "text" value.

- I want to apply a 50% discount to the last cart line item because passes my rule "If message length is less than or equal to 3 characters, the item should get a 50% discount"

However, I can't target the last line item, because they both have the same product variant id "123", so if I create a ProductVariantTarget with "id" = "123", then both line items will be discounted, instead of the last one. I can specify a quantity to make sure only one of them is targeted, but I can't choose which one.

**Comment by pbtripletree** [comment]
Recently tried implementing logic to skip discounting any line item that is a subscription. However, if there are two line items with the same variant, one that isn't a subscription purchase and one that isn't, the discount applies to both as we can only target by variant id 😔 definitely a big blocker for extending our discount function's capabilities!

**Reply by weotch** [reply]
You can kinda work around this by specifying a `quantity` for the `productVariant` target, but it [may not attribute the discount to the right line](https://community.shopify.com/c/shopify-functions/add-a-quot-cartline-quot-target-for-discount-functions/m-p/2204264#M853%3Futm_source=communitymembers&utm_medium=email&utm_campaign=immediate_general).

**Comment by blacklen** [comment]
Is there any updates on this one?

**Comment by gabrielvictoria96** [comment]
I'm facing the SAME issue, good to know there are more people facing it too. I was thinking that was a problem with my code. 

Is there any updates on this?

**Comment by owenfranssen** [comment]
I'm working on a free gift feature. These bugs are seriously affecting what we can do.
I.e. if I add two of the same variant to the cart, one as a gift, the other as a full priced item, the function will discount both regardless
If I add multiple different gifts, the function will only discount one.
This seems illogical when a function can return an object with multiple variant ids each with their own discounts. Why would you not honour that logic?

**Reply by bkspace** [reply]
@owenfranssen - there is also a 'discountApplicationStrategy', so although you are returning multiple discounts, the application strategies available are 'FIRST' or 'MAXIMUM'.

When creating a 'target', you can pass a quantity, such that if you pass a quantity of 1, only 1 quantity of the variant will get the discount.

**Comment by pablogiralt** [comment]
+1

**Comment by jacknavynicy** [comment]
+1
or allowed us to apply the discount with some condition, eg:
product variant that with specific attributes/properties

this will be more useful in our case, if we can have some way to target the specific line item with some condition

**Reply by abhisheksingh75** [reply]
+1..blocker for us..please resolve this

**Reply by owenfranssen** [reply]
+1 here too

**Comment by ct-Marvin** [comment]
+1

**Comment by overallduka** [comment]
+1, that would be useful and makes totally sense.

**Comment by nickwesselman** [comment]
Please use the built-in upvoting instead of "+1" comments y'all. Thank you!

**Reply by jacknavynicy** [reply]
Hi @nickwesselman for this tread, is there any schedule, when will this to be fixed? Thank you!

**Comment by elvonkh** [comment]
I am having an issue with assigning discounts for same variant, but different attributes.

My real case is:
-  l have 3 products in my cart and they have different attributes, I expect function to assign 3 different discounts for 3 different line items (same variant, but different line item properties). I have 3 different discounts for a single variant, which are being assigned according to the Line Item attributes (properties). However, discount application strategy limits restricts me with only one discount.

Also will this feature support assignment of discounts for multiple line items?

Important question. When can we expect the feature to be released?

Is this feature will work along with assigning multiple discounts via single Discount Function? 

Reference:
[Discounts Function Thread](https://github.com/Shopify/function-examples/discussions/194#discussioncomment-7273691)

**Reply by TemurbekRuziyev** [reply]
+1

**Comment by TemurbekRuziyev** [comment]
I'm also facing a problem when it comes to applying discounts to the same product variant but they are different line items
Product discount extension is returning the correct result, but only one of them is  getting a discount

**Comment by roman-yakobnyuk** [comment]
We faced the same problem while investigating a possible migration from Shopify Script to a Function for our Plus client. We had to move back to scripts for now as their use case is not supported. 

Specifically, we are developing "pair" discounts, where every second product of a specific group or type gets a discount, for example, 50%. To achieve that we first need to order products by price from highest to lowest, and then for every second product - apply a discount. 

As the first and second line items can contain the same product - there is currently no way to apply such a discount using a function, but with scripts, we can do it very easily.

**Comment by jonathanmoore** [comment]
The limitation to only target productVariant is a hard blocker for our Shopify Plus client to migrate from Shopify Scripts. 

We have a lot of critical logic tied to campaigns and promos that are not possible with the current state of the Product Discount API. Like everyone else says we need to be able to target line items vs the very broad productVariant.

**Comment by ashfaqhaq** [comment]
Facing the same issue, would really love to know if the team has any plans on releasing this in the upcoming releases.

Thank you!

**Comment by dianacheptene** [comment]
All, this enhancement is currently planned for Q1 of next year.

**Reply by gil--** [reply]
🙏 Thanks for this update. Looking forward to implementing it.

cc @jonathanmoore @pablogiralt @bkspace @avocadoslab

**Reply by didierdonne** [reply]
Hi Diana, is there an update on this?  Thank you.

**Comment by mrpacman101** [comment]
We also face the same issue which prevents us from switching to functions from scripts and also using the 1step checkout. It's a major and critical step to switch to the new checkout but impossible atm with scripts.

Also when scripts is deprecating and support drops end of this year but the new functions is not yet fully usable then will there be an extension of the deadline or what is the official deadline for Shopify scripts? Can you tell when in Q1 because our clients are anxious and need clear timelines? Is it end of Q1 or beginning?

**Comment by futureandahope7** [comment]
This seems to be moving in the right direction. However, I wish to clarify. We need a feature where any line item can be discounted by any amount. That there can be multiple line discounts, not just one. That discounting can be applied in any way we choose.

i.e. we could have a cart:

177 - WPI (chocolate) x 2
178 - WPI (vanilla) X 1
179 - BCAA 7000 (peach) X 3

Within one discount function we would need to be able to apply any kind of discount to any line item.

Our internal rules may be:

```
WPI_combined_quantity = (WPI (chocolate) x 2) + (WPI (vanilla) X 1)

swich(WPI_combined_quantity){
   case 2: 10% discount
   case 3: 20% discount
   case > 3 30% discount
} 

BCAA_combined_quantity = (BCAA 7000 (peach) X 3)

swich(BCAA_combined_quantity){
   case 2: 15% discount
   case 3: 25% discount
   case > 3 35% discount
} 
```

Which would give:

WPI (chocolate) x 2 (20% discount)
WPI (vanilla) X 1 (20% discount)
BCAA 7000 (peach) X 3 (25% discount)

i.e.

```
// assigned from case conditions
BCAA_discount = '25%';
WPI_discount = '20%'

targetsWPI = [{
      // Use the variant ID to create a discount target
      cartLineItem: {
        id: 177
      }
    }, 
{
      // Use the variant ID to create a discount target
      cartLineItem: {
        id: 178
      }
    }
];

targetsBCAA = [{
      // Use the variant ID to create a discount target
      cartLineItem: {
        id: 179
      }
    }
]

return {
    discounts: [
      {
        // Apply the discount to the collected targets
        targetsWPI,
        // Define a percentage-based discount
        value: {
          percentage: {
            value: WPI_discount
          }
        }
      },
 {
        // Apply the discount to the collected targets
        targetsBCAA,
        // Define a percentage-based discount
        value: {
          percentage: {
            value: BCAA_discount
          }
        }
      }
    ],
    discountApplicationStrategy: DiscountApplicationStrategy.First
  };
```

**Reply by AndreaMorone** [reply]
At the moment is possible to apply all discounts at once with one function using DiscountApplicationStrategy.ALL in API 2024-01.

Instead to target discount by line item ID we need to wait a little bit, Shopify has it in Q1 2024

**Comment by evaldas-raisutis** [comment]
Couldn't we use the cart transform API to set line item prices? https://shopify.dev/changelog/new-and-updated-operations-for-the-cart-transform-api

Has anyone looked into this?

**Reply by ct-Marvin** [reply]
You could change the actual line item price using the newly released change operation. This has two problems though:

1. All Cart transform discounts, do not apply an actual discount code so the Discount is not visible in the cart/checkout like normal discounts, you only see the updated price
2. The change operation can only be used by Shopify Plus stores

In case you are considering using the expand operation, this lets you set the price of the elements inside a bundle, not the price of the bundle's parent or if you use a percentage discount, it does change the parent's price but also the bundle component's price based on a "weighted price algorythm".

**Comment by elvonkh** [comment]
I just checked `2024-01 latest` and `2024-04 release candidate` and I don't see any changes in Product Discount API. When can we expect this feature to be released? This is causing a lot of bugs.

**Reply by gil--** [reply]
Roadmap shows Q3 (July - September) https://shopify.dev/docs/apps/checkout#shopify-functions

**Comment by rgilbert82** [comment]
+1
I would also like to know the timeline when this feature will be released for Discounts API. Thanks.

**Reply by gil--** [reply]
Roadmap shows Q3 (July - September) https://shopify.dev/docs/apps/checkout#shopify-functions

**Reply by fxchain** [reply]
+1
Hope they roll this out before they disable Shopify Scripts.

**Comment by thosakwe** [comment]
This causes problems when discounts are applied to a cart where both a subscription and one-time purchase version of the same product variant are in the cart.

Because discounts target only variant IDs, both will receive the discount.

Most merchants I've spoken to were trying to create setups where only the one-time purchase version would get discounted.

**Comment by dpwolf** [comment]
Hey folks, we've got good news for you!

As of the `2024-07` API version the product discount function enables [cart line targets](https://shopify.dev/docs/api/functions/reference/product-discounts/graphql/common-objects/target?api%5Bversion%5D=2024-07)!

**Comment by nickwesselman** [comment]
✅

**Comment by realpratiknikam** [comment]
Works like a charm! Thanks @dpwolf @nickwesselman 🔥
